Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1332,6 +1332,62 @@ The following approach can be used in order to patch the JavaScript file:
 5. Put the *patched code* on a single line and copy it in the original `assets/index.android.bundle` file.
 6. Repack the APK archive using `apktool` tool and sign it before to install it on the target device/emulator.
 
+##### Library Injection
+
+In the previous section we learned about patching application code to assist in our analysis. Patching an application in theory can be used to perform any task, but in practice it has its own limitations. For example, we have a 3rd party application for which we don't have the source code and only works locally. We want to modify this application so that it is accessible over the network as well. One possible way is to patch the application's APIs and introduce networking code, but this approach may not be practical for a large application. Another shortcoming being the patching performed is not reusable for any other executable, i.e patching performed is unique to each application. As an alternative approach, if we can separately develop and compile a library which provides the networking logic and inject this into the application, not only this will be time efficient, but also the library will be reusable. This paradigm of introducing a separately compiled code in a form of a library in a process memory is often referred to as library injection. In Windows world this technique has been in use for a long time, and is referred as *DLL injection*. It is commonly used to modify and bypass anti-cheat mechanisms in games. On Android and iOS, most common example is Frida, under-the-hood it is performing a library injection to provide all the functionalities.
```

Comment:
Split in 2 paragraphs: first motivation/limitations, second intro/description. The Frida example use case is a bit more accurate.
```suggestion
In the previous section we learned about patching application code to assist in our analysis, but this approach has several limitations. For instance, you'd like to log everything that's being sent over the network without having to perform a MITM attack. For this you'd have to patch all possible calls to the network APIs, which can quickly become impracticable when dealing with large applications. In addition, the fact that patching is unique to each application can be also considered a shortcoming, as this code cannot be easily reused.

Using library injection you can develop reusable libraries and inject them to different applications, effectively making them behave differently without having to modify their original source code. This is known as DLL injection on Windows (broadly used to modify and bypass anti-cheat mechanisms in games), `LD_PRELOAD` on Linux and `DYLD_INSERT_LIBRARIES` on macOS. On Android and iOS, a common example is using the Frida Gadget whenever Frida's so-called [Injected mode](https://frida.re/docs/modes/#injected "Frida Injected Mode") of operation isn’t suitable (i.e. you cannot run frida-server on the target device). In this situation, you can [inject the Gadget](https://frida.re/docs/gadget/ "Frida Gadget") library by using the same methods you're going to learn in this section.
```